src/units/abilities.cpp

```diff
@@ -1877,21 +1902,29 @@ bool attack_type::special_active_impl(
 	const config& special_backstab = special["backstab"].to_bool() ? cfg : special;
 
 	// Filter the units involved.
-	//If filter concerns the unit on which special is applied,
-	//then the type of special must be entered to avoid calling
-	//the function of this special in matches_filter()
-	std::string self_tag_name = whom_is_self ? tag_name : "";
-	if (!special_unit_matches(self, other, self_loc, self_attack, special, is_for_listing, filter_self, self_tag_name))
-		return false;
-	std::string opp_tag_name = !whom_is_self ? tag_name : "";
-	if (!special_unit_matches(other, self, other_loc, other_attack, special_backstab, is_for_listing, "filter_opponent", opp_tag_name))
-		return false;
-	std::string att_tag_name = is_attacker ? tag_name : "";
-	if (!special_unit_matches(att, def, att_loc, att_weapon, special, is_for_listing, "filter_attacker", att_tag_name))
-		return false;
-	std::string def_tag_name = !is_attacker ? tag_name : "";
-	if (!special_unit_matches(def, att, def_loc, def_weapon, special, is_for_listing, "filter_defender", def_tag_name))
-		return false;
+	// Use int variables for have an limit to infinite recursion and prevent a crash.
+	int num_rec = self_attack->num_recursion;
+	if(num_rec < 3){
+		if (!special_unit_matches(self, other, self_loc, self_attack, special, is_for_listing, filter_self, num_rec)){
+			self_attack->num_recursion = 0;
+			return false;
+		}
+		if (!special_unit_matches(other, self, other_loc, other_attack, special_backstab, is_for_listing, "filter_opponent", num_rec)){
+			self_attack->num_recursion = 0;
+			return false;
+		}
+		if (!special_unit_matches(att, def, att_loc, att_weapon, special, is_for_listing, "filter_attacker", num_rec)){
+			self_attack->num_recursion = 0;
+			return false;
+		}
+		if (!special_unit_matches(def, att, def_loc, def_weapon, special, is_for_listing, "filter_defender", num_rec)){
+			self_attack->num_recursion = 0;
+			return false;
+		}
+	}
+	//when exit of the loop, reinitialise variables
+	self_attack->num_recursion = 0;
+	self_attack->global_num_recursion = 0;
```

Comment:
I think a global `int` is going to end up being the solution, but instead of having lots of ways to reset it, it should have a RAII class (increment in the constructor, decrement in the destructor).

Reply:
i rebased locally with https://github.com/wesnoth/wesnoth/pull/8977/commits/906f16e51ab2360d7a2eabce3929c749c921b1cb and it's work same without global_num_recursion. Strange.

Reply:
and it's infinite loop in wml test yet, i don't understand

Reply:
I'll have to write a long comment about what I mean by RAII, but that will need to wait until next week.

Reply:
write 'RAII' in all characters, and i perhaps unerstand by myself.

Reply:
i read docs on RAII and i see thatwhat you want said. I do choose the AIII to use and encoding that properly

Reply:
@stevecotton int this context i don't know how use porperly RAII class because it is after each checking what variables must be reinitialized